src/main/java/io/cdap/plugin/gcp/gcs/connector/GCSConnector.java

```diff
@@ -0,0 +1,234 @@
+import io.cdap.cdap.etl.api.connector.BrowseDetail;
+import io.cdap.cdap.etl.api.connector.BrowseEntity;
+import io.cdap.cdap.etl.api.connector.BrowseEntityProperty;
+import io.cdap.cdap.etl.api.connector.BrowseRequest;
+import io.cdap.cdap.etl.api.connector.Connector;
+import io.cdap.cdap.etl.api.connector.ConnectorContext;
+import io.cdap.cdap.etl.api.connector.ConnectorSpec;
+import io.cdap.cdap.etl.api.connector.ConnectorSpecRequest;
+import io.cdap.cdap.etl.api.connector.PluginSpec;
+import io.cdap.cdap.etl.api.validation.ValidationException;
+import io.cdap.plugin.format.connector.AbstractFileConnector;
+import io.cdap.plugin.format.connector.FileTypeDetector;
+import io.cdap.plugin.gcp.common.GCPConnectorConfig;
+import io.cdap.plugin.gcp.common.GCPUtils;
+import io.cdap.plugin.gcp.gcs.GCSPath;
+import io.cdap.plugin.gcp.gcs.source.GCSSource;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * GCS Connector
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(GCSConnector.NAME)
+@Category("Google Cloud Platform")
+@Description("Connector to browse and sample from Google Cloud Storage")
+public class GCSConnector extends AbstractFileConnector<GCPConnectorConfig> {
+  public static final String NAME = "GCS";
+  static final String BUCKET_TYPE = "bucket";
+  static final String DIRECTORY_TYPE = "directory";
+  static final String FILE_TYPE = "file";
+  static final String LAST_MODIFIED_KEY = "Last Modified";
+  static final String SIZE_KEY = "Size";
+  static final String FILE_TYPE_KEY = "File Type";
+
+  private GCPConnectorConfig config;
+
+  public GCSConnector(GCPConnectorConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @Override
+  public void test(ConnectorContext context) throws ValidationException {
+    String project = config.tryGetProject();
+    FailureCollector failureCollector = context.getFailureCollector();
+    if (project == null) {
+      failureCollector
+        .addFailure("Could not detect Google Cloud project id from the environment.", "Please specify a project id.")
+        .withConfigProperty(GCPConnectorConfig.NAME_PROJECT);
+    }
+
+    Boolean isServiceAccountFilePath = config.isServiceAccountFilePath();
+    if (isServiceAccountFilePath == null) {
+      failureCollector.addFailure("Service account type is undefined.", "Must be `filePath` or `JSON`");
+    }
+
+    // no need to continue here as we are not able to continue validating
+    if (!failureCollector.getValidationFailures().isEmpty()) {
+      return;
+    }
+
+    Credentials credentials = null;
+    try {
+      credentials =
+        GCPUtils.loadServiceAccountCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+    } catch (IOException e) {
+      failureCollector.addFailure(String.format("Service account provided is not valid: %s.", e.getMessage()),
```

Comment:
would be good to add the stack trace to the failure here and below, as the exception message may not be enough to figure out what is wrong.